File: frontend/src/api/ecr.ts
import api from './index'

export interface ECRItem {
  id: number
  code: string
  title: string
  ecr_type: string
  urgency: string
  status: string
  submitter_name?: string
  created_at: string
}

export interface ECRDetail extends ECRItem {
  reason: string
  affected_products?: Record<string, unknown>
  affected_documents?: Record<string, unknown>
  description?: string
  submitter_id: number
  reviewer_id?: number
  rejection_reason?: string
  reviewed_at?: string
  updated_at: string
  attachments: Array<{
    id: number
    file_name: string
    file_type?: string
    file_size: number
    created_at: string
  }>
}

export const ecrApi = {
  list(params?: { skip?: number; limit?: number; status?: string; ecr_type?: string }) {
    return api.get<ECRItem[]>('/ecr/', { params })
  },

  get(id: number) {
    return api.get<ECRDetail>(`/ecr/${id}`)
  },

  create(data: {
    title: string
    ecr_type?: string
    reason: string
    urgency?: string
    description?: string
  }) {
    return api.post<ECRDetail>('/ecr/', data)
  },

  update(id: number, data: Record<string, unknown>) {
    return api.put<ECRDetail>(`/ecr/${id}`, data)
  },

  submit(id: number) {
    return api.post<ECRDetail>(`/ecr/${id}/submit`)
  },

  review(id: number, data: { action: string; rejection_reason?: string }) {
    return api.post<ECRDetail>(`/ecr/${id}/review`, data)
  },
}

File: backend/app/api/v1/ecr.py
"""Phase 2 — ECR API: 工程变更申请接口"""

import logging
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status, Request
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.core.security import get_current_user
from app.models.user import User
from app.crud.ecr import ecr_crud
from app.schemas.ecr_eco import (
    ECRCreate, ECRUpdate, ECRRejectRequest, ECRReviewRequest,
    ECRSummaryOut, ECRDetailOut,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/ecr", tags=["ECR"])


@router.post("/", response_model=ECRDetailOut, status_code=status.HTTP_201_CREATED)
async def create_ecr(
    data: ECRCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """创建 ECR"""
    ecr = await ecr_crud.create(
        db,
        data=data.model_dump(),
        submitter_id=current_user.id,
        submitter_name=current_user.username,
    )
    await db.commit()
    return ecr


@router.get("/", response_model=list[ECRSummaryOut])
async def list_ecrs(
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    status: Optional[str] = None,
    ecr_type: Optional[str] = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """列出 ECR"""
    return await ecr_crud.list(
        db, skip=skip, limit=limit, status=status, ecr_type=ecr_type
    )


@router.get("/{ecr_id}", response_model=ECRDetailOut)
async def get_ecr(
    ecr_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """获取 ECR 详情"""
    ecr = await ecr_crud.get(db, ecr_id)
    if not ecr:
        raise HTTPException(status_code=404, detail="ECR 不存在")
    return ecr


@router.put("/{ecr_id}", response_model=ECRDetailOut)
async def update_ecr(
    ecr_id: int,
    data: ECRUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """编辑 ECR（仅 draft 状态）"""
    ecr = await ecr_crud.get(db, ecr_id)
    if not ecr:
        raise HTTPException(status_code=404, detail="ECR 不存在")
    try:
        ecr = await ecr_crud.update(db, ecr, data.model_dump(exclude_unset=True))
        await db.commit()
        return ecr
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{ecr_id}/submit", response_model=ECRDetailOut)
async def submit_ecr(
    ecr_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """提交 ECR: DRAFT → SUBMITTED"""
    ecr = await ecr_crud.get(db, ecr_id)
    if not ecr:
        raise HTTPException(status_code=404, detail="ECR 不存在")
    try:
        ecr = await ecr_crud.submit(db, ecr)
        await db.commit()
        return ecr
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{ecr_id}/review", response_model=ECRDetailOut)
async def review_ecr(
    ecr_id: int,
    data: ECRReviewRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """审批 ECR: SUBMITTED → APPROVED/REJECTED"""
    ecr = await ecr_crud.get(db, ecr_id)
    if not ecr:
        raise HTTPException(status_code=404, detail="ECR 不存在")
    try:
        ecr = await ecr_crud.review(
            db, ecr,
            action=data.action,
            reviewer_id=current_user.id,
            rejection_reason=data.rejection_reason,
        )
        await db.commit()
        return ecr
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
